ambry-store/src/main/java/com.github.ambry.store/BlobStore.java

```diff
@@ -408,9 +416,16 @@ public void put(MessageWriteSet messageSetToWrite) throws StoreException {
           logger.trace("All entries were absent, and were written to the store successfully");
           break;
       }
+      onSuccess();
     } catch (StoreException e) {
+      if (e.getErrorCode() == StoreErrorCodes.IOError) {
```

Comment:
1. This PR ensures that `StoreException.IOError` is only used for real disk I/O errors. `IOException` has a broader domain and some of it are not related to disk hardware issue.
2. `StoreException.IOError` is not always triggered by `IOException`, also it could be an `InternalError`. Following is an example:
```   
   } catch (InternalError e) {
      throw e.getMessage().equals(StoreException.INTERNAL_ERROR_STR) ? new StoreException(
          "Internal error occurred due to unsafe memory access", e, StoreErrorCodes.IOError)
          : new StoreException("Unknown internal error while trying to get store key", e,
              StoreErrorCodes.Unknown_Error);
    } catch (IOException e) {
      throw e.getMessage().equals(StoreException.IO_ERROR_STR) ? new StoreException(
          "IO error while trying to get store key", e, StoreErrorCodes.IOError)
          : new StoreException("Unknown IO error while trying to get store key", e, StoreErrorCodes.Unknown_Error);
    } catch (Throwable t) {
      throw new StoreException("Unknown error while trying to get store key ", t, StoreErrorCodes.Unknown_Error);
    }
```

Reply:
Right - so I am just looking at whether both these exceptions should be handled for disk errors. Is it possible to not have to handle `IOException` here at all? As in, does it make sense (and is it possible) to convert all `IOException` that are actual disk errors as `StoreException.IOError`?

Reply:
Is `DiskIOError` a better name as it has small scope?